go/cmd/dolt/commands/log_graph.go

```diff
@@ -0,0 +1,428 @@
+// if the commit is a merge commit, add one more line for the "Merge:" line
+func getHeightOfCommit(commit *CommitInfoWithChildren) int {
+	height := 4 + len(commit.formattedMessage)
+	if len(commit.Commit.parentHashes) > 1 {
+		height = height + 1
+	}
+	return height
+}
+
+// print the commit messages in the graph matrix
+func appendMessage(graph [][]string, pager *outputpager.Pager, apr *argparser.ArgParseResults, commits []*CommitInfoWithChildren) {
+	decoration := apr.GetValueOrDefault(cli.DecorateFlag, "auto")
+
+	for i := 0; i < len(commits)-1; i++ {
+		startY := commits[i].Y
+		endY := commits[i+1].Y
+		startX := commits[i].X + 1
+
+		// find the maximum x position of the graph in the range startY to endY
+		// this is used to align the commit message with the graph without overlapping with the graph
+		for j := startY; j < endY; j++ {
+			if len(graph[j]) > startX {
+				startX = len(graph[j])
+			}
+		}
+
+		printCommitMetadata(graph, pager, startY, startX, commits[i], decoration)
+
+		commitInfoHeight := getHeightOfCommit(commits[i])
+
+		// print the graph with commit message
+		for j, line := range commits[i].formattedMessage {
+			y := startY + commitInfoHeight - len(commits[i].formattedMessage) + j
+			printLine(graph, startX, y, pager, fmt.Sprintf("\t%s", line), commits[i].Commit, "\033[37m", "no")
+		}
+
+		// print the remaining lines of the graph of the current commit
+		for j := startY + commitInfoHeight; j < endY; j++ {
+			pager.Writer.Write([]byte(strings.Join(graph[j], "")))
+			pager.Writer.Write([]byte("\n"))
+		}
+	}
+
+	last_commit_y := commits[len(commits)-1].Y
+	printCommitMetadata(graph, pager, last_commit_y, len(graph[last_commit_y]), commits[len(commits)-1], decoration)
+	for _, line := range commits[len(commits)-1].formattedMessage {
+		pager.Writer.Write([]byte(fmt.Sprintf("\t\033[37m%s", line)))
+		pager.Writer.Write([]byte("\n"))
+	}
+}
+
+// expand the graph based on the length of the commit message
+func expandGraph(commits []*CommitInfoWithChildren, width int) {
+	posY := 0
+	for i, commit := range commits {
+		// one empty column between each branch path
+		commit.X = commit.X * 2
+		commit.Y = posY
+		formattedMessage := wrapTextOnWidth(commit.Commit.commitMeta.Description, width)
+		commit.formattedMessage = formattedMessage
+
+		// make sure there is enough space for the diagonal line connecting to the parent
+		// this is an approximation, assume that there will be enough space if parent is not the next commit
+		maxDistanceFromParent := float64(0)
+		if i < len(commits)-1 && commits[i+1].Commit.commitHash == commit.Commit.parentHashes[0] {
+			maxDistanceFromParent = math.Max(math.Abs(float64(commits[i+1].X-commit.X)), maxDistanceFromParent)
+		}
+
+		commitInfoHeight := getHeightOfCommit(commit)
+		posY += int(math.Max(float64(commitInfoHeight+1), maxDistanceFromParent))
+	}
+}
+
+func logGraph(pager *outputpager.Pager, apr *argparser.ArgParseResults, commitInfos []CommitInfo) {
```

Comment:
looks like the apr instance is only used to pass through the decoration flag. It would be better to use a more simple type.

Reply:
I'm thinking adding more support in the next PR, such as `dolt log --graph --oneline`.